internal/icinga2/client_api.go

```diff
@@ -275,17 +305,17 @@ func (client *Client) checkMissedChanges(ctx context.Context, objType string, ca
 			if err != nil {
 				return fmt.Errorf("failed to construct Event from Acknowledgement response, %w", err)
 			}
-		} else if isMuted(&objQueriesResult) {
+		} else if checkableIsMuted {
 			fakeEv, err = client.buildCommonEvent(ctx, hostName, serviceName)
 			if err != nil {
 				return fmt.Errorf("failed to construct checkable fake mute event: %w", err)
 			}
 
 			fakeEv.Type = event.TypeMute
-			if attrs.IsFlapping {
-				fakeEv.SetMute(true, "Checkable is flapping, but we missed the Icinga 2 FlappingStart event")
-			} else {
+			if attrs.DowntimeDepth != 0 {
```

Comment:
Could you please add a another note to `HostServiceRuntimeAttributes`' API documentation stating the relationship between `DowntimeDepth` and flapping.

Reply:
Like what for example? I'm not sure what I'm supposed to document here, but there is no relationship between those fields. I have switched the condition here because the conditions in the `isMuted()` function are also switched, so that when this function returns `true` because `downtimedepth != 0`, it doesn't go into the `flapping` branch.

Reply:
I don't know and that's why I asked about it. When reading the current code, I am surprised that `DowntimeDepth` decides if the object is either in a downtime or might be flapping. As there are three possibilities leading to a muted event, a bit more prose would at least help me understanding what's going on.

Maybe my suggested place is wrong and the `isMuted` function could get some additional documentation stating the implications or even get multiple returns like, e.g., `(isMutedAck, isMutedDowntime, isMutedFlapping bool, err error)` to remove domain-knowledge later on, like at this very place.

Reply:
> As there are three possibilities leading to a muted event, a bit more prose would at least help me understanding what's going on.

Right, there are three possibilities and the first case is evaluated here.
https://github.com/Icinga/icinga-notifications/blob/03b8954092abccb320957e9e3f30f56560116a43/internal/icinga2/client_api.go#L294

The second one:
https://github.com/Icinga/icinga-notifications/blob/03b8954092abccb320957e9e3f30f56560116a43/internal/icinga2/client_api.go#L315

When those first two cases yield `false`, only one case remains, and that is `flapping` and has nothing to do with the above downtime check.
https://github.com/Icinga/icinga-notifications/blob/03b8954092abccb320957e9e3f30f56560116a43/internal/icinga2/client_api.go#L317-L319

Reply:
My point was that it adds additional mental load on the caller. This domain knowledge is necessary both within `isMuted` and outside after being called. Thus, there is no clean barrier and I have to exactly know how muting works, even outside the `isMuted` function.

For me, not having the amount of domain specific knowledge you do, I had to jump around the code and the Icinga 2 documentation to actually understand how it works. If I'm going to look at this next month, I have to recreate this mental model. That's all I wanted to point out. Technically this should work, but when I'm looking at the code the next time, I don't exactly know why.

Reply:
> Thus, there is no clean barrier and I have to exactly know how muting works, even outside the isMuted function.

Sorry, but more than commenting on when `isMuted()` returns true and when it doesn't, I don't really know what else I should do to make it clearer for you. I haven't even changed the existing code that much that could confuse you even more, I have literally just switched that single condition.

> or even get multiple returns like, e.g., `(isMutedAck, isMutedDowntime, isMutedFlapping bool, err error)` to remove domain-knowledge later on, like at this very place.

I don't see how that is supposed to make things any better, the `isMuted` utility is simply meant to return true or false based on the given conditions, and if you want it to return something like that, I don't know why you would need such utility then.

> Maybe my suggested place is wrong and the isMuted function could get some additional documentation stating the implications

Sorry, but I failed to get what implications you are referring to. As I said above, there are no implications, this utility simply returns true or false based on a set condition and I'm checking them here one after the other to generate more specific events as to why an event/object is being muted. If you're having a hard time understanding the general handling of incident muting, we can have a call and discuss this in person, or I may be completely misunderstanding you and @julianbrost could just jump in and put a stop to it if he understands your point better.

Reply:
Sorry if my prior rambling wasn't on point. In a nutshell, I wanted to address a missing [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). Please take a look at the following diff where I tried to change the signature of `isMuted`, resulting in removing the logic from `checkMissedChanges`.


```diff
diff --git a/internal/icinga2/client.go b/internal/icinga2/client.go
index 82cc20a..a30152b 100644
--- a/internal/icinga2/client.go
+++ b/internal/icinga2/client.go
@@ -265,7 +265,7 @@ func (client *Client) buildAcknowledgementEvent(ctx context.Context, ack *Acknow
 		if err != nil {
 			return nil, err
 		}
-		if muted, err := isMuted(ctx, client, queryResult); err != nil {
+		if muted, _, _, _, err := isMuted(ctx, client, queryResult); err != nil {
 			return nil, err
 		} else if !muted {
 			ev.Message = queryResult.Attrs.LastCheckResult.Output
@@ -312,7 +312,7 @@ func (client *Client) buildDowntimeEvent(ctx context.Context, d Downtime, startE
 		if err != nil {
 			return nil, err
 		}
-		if muted, err := isMuted(ctx, client, queryResult); err != nil {
+		if muted, _, _, _, err := isMuted(ctx, client, queryResult); err != nil {
 			return nil, err
 		} else if !muted {
 			// When a downtime is cancelled/expired and there's no other active downtime/ack, we're going to send some
@@ -351,7 +351,7 @@ func (client *Client) buildFlappingEvent(ctx context.Context, flapping *Flapping
 		if err != nil {
 			return nil, err
 		}
-		if muted, err := isMuted(ctx, client, queryResult); err != nil {
+		if muted, _, _, _, err := isMuted(ctx, client, queryResult); err != nil {
 			return nil, err
 		} else if !muted {
 			ev.Message = queryResult.Attrs.LastCheckResult.Output
diff --git a/internal/icinga2/client_api.go b/internal/icinga2/client_api.go
index a6cc385..9d85991 100644
--- a/internal/icinga2/client_api.go
+++ b/internal/icinga2/client_api.go
@@ -285,13 +285,13 @@ func (client *Client) checkMissedChanges(ctx context.Context, objType string, ca
 		}
 
 		attrs := objQueriesResult.Attrs
-		checkableIsMuted, err := isMuted(ctx, client, &objQueriesResult)
+		muted, mutedAck, mutedDowntime, mutedFlapping, err := isMuted(ctx, client, &objQueriesResult)
 		if err != nil {
 			return err
 		}
 
 		var fakeEv *event.Event
-		if attrs.Acknowledgement != AcknowledgementNone {
+		if mutedAck {
 			ackComment, err := client.fetchAcknowledgementComment(ctx, hostName, serviceName, attrs.AcknowledgementLastChange.Time())
 			if err != nil {
 				return fmt.Errorf("fetching acknowledgement comment for %q failed, %w", objectName, err)
@@ -305,17 +305,19 @@ func (client *Client) checkMissedChanges(ctx context.Context, objType string, ca
 			if err != nil {
 				return fmt.Errorf("failed to construct Event from Acknowledgement response, %w", err)
 			}
-		} else if checkableIsMuted {
+		} else if muted {
 			fakeEv, err = client.buildCommonEvent(ctx, hostName, serviceName)
 			if err != nil {
 				return fmt.Errorf("failed to construct checkable fake mute event: %w", err)
 			}
 
 			fakeEv.Type = event.TypeMute
-			if attrs.DowntimeDepth != 0 {
+			if mutedDowntime {
 				fakeEv.SetMute(true, "Checkable is in downtime, but we missed the Icinga 2 DowntimeStart event")
-			} else {
+			} else if mutedFlapping {
 				fakeEv.SetMute(true, "Checkable is flapping, but we missed the Icinga 2 FlappingStart event")
+			} else {
+				client.Logger.Error("lolwat")
 			}
 		} else {
 			// This could potentially produce numerous superfluous database (event table) entries if we generate such
diff --git a/internal/icinga2/util.go b/internal/icinga2/util.go
index 440f028..110359b 100644
--- a/internal/icinga2/util.go
+++ b/internal/icinga2/util.go
@@ -24,19 +24,30 @@ func rawurlencode(s string) string {
 // enabled, this will always return false.
 //
 // Returns an error if it fails to query the status of IcingaApplication from the /v1/status endpoint.
-func isMuted(ctx context.Context, client *Client, checkable *ObjectQueriesResult[HostServiceRuntimeAttributes]) (bool, error) {
-	if checkable.Attrs.Acknowledgement != AcknowledgementNone || checkable.Attrs.DowntimeDepth != 0 {
-		return true, nil
+func isMuted(ctx context.Context, client *Client, checkable *ObjectQueriesResult[HostServiceRuntimeAttributes]) (
+	muted, mutedAck, mutedDowntime, mutedFlapping bool, err error,
+) {
+	defer func() { muted = err == nil && (mutedAck || mutedDowntime || mutedFlapping) }()
+
+	if checkable.Attrs.Acknowledgement != AcknowledgementNone {
+		mutedAck = true
+		return
+	}
+	if checkable.Attrs.DowntimeDepth != 0 {
+		mutedDowntime = true
+		return
 	}
 
 	if checkable.Attrs.IsFlapping && checkable.Attrs.EnableFlapping {
-		status, err := client.fetchIcingaAppStatus(ctx)
+		var status *IcingaAppStatus
+		status, err = client.fetchIcingaAppStatus(ctx)
 		if err != nil {
-			return false, err
+			return
 		}
 
-		return status.App.EnableFlapping, nil
+		mutedFlapping = status.App.EnableFlapping
+		return
 	}
 
-	return false, nil
+	return
 }
```

Reply:
As discussed offline, we came to the conclusion of using a similar version as before, but hinting the reader by introducing `checkableIsMuted` to the first if clause.